src/data/pfun/fix.lean

```diff
@@ -0,0 +1,319 @@
+    cases h' with i h',
+    cases hh _ _ h' }
+end
+
+-- lemma approx_eq {i j : ℕ} {a : α} (hi : (approx f i a).dom) (hj : (approx f j a).dom) :
+--   approx f i a = approx f j a :=
+-- begin
+--   simp [dom_iff_mem] at hi hj,
+--   cases hi with b hi, cases hj with b' hj,
+--   have : b' = b,
+--   { wlog hij : i ≤ j := le_total i j using [i j b b',j i b' b],
+--     apply mem_unique hj, apply approx_mono hf hij _ _ hi, },
+--   subst b',
+--   ext y, split; intro hy,
+--   rename hi hh, rename hj hi, rename hh hj,
+--   all_goals
+--   { have := mem_unique hy hj, subst this,
+--     exact hi }
+-- end
+
+noncomputable def approx_chain : chain (Π a, roption $ β a) :=
+begin
+  refine ⟨ approx f, _ ⟩,
+  apply approx_mono, exact hf
+end
+
+lemma le_f_of_mem_approx {x} (hx : x ∈ approx_chain hf) : x ≤ f x :=
+begin
+  revert hx, simp [approx_chain,stream.mem_def],
+  intros i hx, subst x,
+  apply approx_mono' hf
+end
+
+lemma f_mem_approx_chain {x} (hx : x ∈ approx_chain hf) : f x ∈ approx_chain hf :=
+begin
+  revert hx, simp [approx_chain,stream.mem_def],
+  intros i hx, subst hx, exact ⟨i.succ,rfl⟩
+end
+
+lemma approx_mem_approx_chain {i} : approx f i ∈ approx_chain hf :=
+stream.mem_of_nth_eq rfl
+
+end fix
+
+namespace pi
+
+open fix
+
+variables {α} {β : α → Type*}
+open complete_partial_order
+
+variables {f : (Π a, roption $ β a) → (Π a, roption $ β a)} (hf : monotone f)
+
+include hf
+
+open roption (hiding Sup) nat
+open nat.up complete_partial_order
+
+lemma fix_eq_Sup : fix f = Sup (approx_chain hf) :=
+begin
+  apply le_antisymm,
+  { intro x, cases min_fix hf x with i hx,
+    transitivity' approx f i.succ x,
+    { transitivity', apply hx, apply approx_mono' hf },
+    apply le_Sup _ _ _, dsimp [approx],
+    rw chain.mem_map_iff,
+    refine ⟨approx f i.succ,_,rfl⟩,
+    apply approx_mem_approx_chain },
+  { apply Sup_le _ _ _,
+    simp [mem_map_iff,approx_chain,stream.mem_def],
+    intros y x, revert y, simp, apply max_fix hf },
+end
+
+@[main_declaration]
+lemma fix_le {X : Π a, roption $ β a} (hX : f X ≤ X) : fix f ≤ X :=
+begin
+  rw fix_eq_Sup hf,
+  apply Sup_le _ _ _,
+  simp [approx_chain,stream.mem_def,stream.nth],
+  intros y i, revert y, simp,
+  induction i, apply bot_le _,
+  transitivity' f X, apply hf i_ih,
+  apply hX
+end
+
+variables {hf} (hc : continuous f hf)
+include hc
+
+lemma fix_eq : fix f = f (fix f) :=
+begin
+  rw [fix_eq_Sup hf,hc],
+  apply le_antisymm,
+  { apply Sup_le_Sup_of_le _,
+    intros x hx, existsi [f x,chain.mem_map _ hf _ hx],
+    apply le_f_of_mem_approx _ hx },
+  { apply Sup_le_Sup_of_le _,
+    intros x hx, rw chain.mem_map_iff at hx,
+    rcases hx with ⟨y,h₀,h₁⟩, refine ⟨x,_,le_refl _⟩,
+    rw ← h₁, apply f_mem_approx_chain _ h₀ }
+end
+
+end pi
+
+class has_fix (α : Type*) :=
+(fix : (α → α) → α)
+
+open has_fix complete_partial_order
+
+class lawful_fix (α : Type*) [has_fix α] [complete_partial_order α] :=
+(fix_eq : ∀ {f : α → α}, continuous' f → fix f = f (fix f))
```

Comment:
This (and most of the rest of the file) looks like it belongs in `control` somewhere

Reply:
I can move the file.